gateway/gateway_test.go

```diff
@@ -496,325 +102,660 @@ func TestPretty404(t *testing.T) {
 		{"/nope", "text/html", http.StatusNotFound, "Custom 404"},
 		{"/nope", "text/*", http.StatusNotFound, "Custom 404"},
 		{"/nope", "*/*", http.StatusNotFound, "Custom 404"},
-		{"/nope", "application/json", http.StatusNotFound, fmt.Sprintf("failed to resolve /ipns/example.net/nope: no link named \"nope\" under %s\n", k.Cid().String())},
+		{"/nope", "application/json", http.StatusNotFound, fmt.Sprintf("failed to resolve /ipns/example.net/nope: no link named \"nope\" under %s\n", root.String())},
 		{"/deeper/nope", "text/html", http.StatusNotFound, "Deep custom 404"},
 		{"/deeper/", "text/html", http.StatusOK, ""},
 		{"/deeper", "text/html", http.StatusOK, ""},
 		{"/nope/nope", "text/html", http.StatusNotFound, "Custom 404"},
 	} {
 		testName := fmt.Sprintf("%s %s", test.path, test.accept)
 		t.Run(testName, func(t *testing.T) {
-			var c http.Client
-			req, err := http.NewRequest("GET", ts.URL+test.path, nil)
-			assert.NoError(t, err)
+			req := mustNewRequest(t, "GET", ts.URL+test.path, nil)
 			req.Header.Add("Accept", test.accept)
 			req.Host = host
-			resp, err := c.Do(req)
-			assert.NoError(t, err)
+			resp := mustDo(t, req)
 			defer resp.Body.Close()
-			assert.Equal(t, test.status, resp.StatusCode)
+			require.Equal(t, test.status, resp.StatusCode)
 			body, err := io.ReadAll(resp.Body)
-			assert.NoError(t, err)
+			require.NoError(t, err)
 			if test.text != "" {
-				assert.Equal(t, test.text, string(body))
+				require.Equal(t, test.text, string(body))
 			}
 		})
 	}
 }
 
-func TestBrowserErrorHTML(t *testing.T) {
-	ts, _, root := newTestServerAndNode(t, nil)
-	t.Logf("test server url: %s", ts.URL)
+func TestHeaders(t *testing.T) {
+	t.Parallel()
 
-	t.Run("plain error if request does not have Accept: text/html", func(t *testing.T) {
-		t.Parallel()
+	ts, _, _ := newTestServerAndNode(t, nil, "headers-test.car")
 
-		req, err := http.NewRequest(http.MethodGet, ts.URL+"/ipfs/"+root.String()+"/nonexisting-link", nil)
-		assert.Nil(t, err)
+	var (
+		rootCID = "bafybeidbcy4u6y55gsemlubd64zk53xoxs73ifd6rieejxcr7xy46mjvky"
 
-		res, err := doWithoutRedirect(req)
-		assert.Nil(t, err)
-		assert.Equal(t, http.StatusNotFound, res.StatusCode)
-		assert.NotContains(t, res.Header.Get("Content-Type"), "text/html")
+		dirCID   = "bafybeihta5xfgxcmyxyq6druvidc7es6ogffdd6zel22l3y4wddju5xxsu"
+		dirPath  = "/ipfs/" + rootCID + "/subdir/"
+		dirRoots = rootCID + "," + dirCID
 
-		body, err := io.ReadAll(res.Body)
-		assert.Nil(t, err)
-		assert.NotContains(t, string(body), "<!DOCTYPE html>")
-	})
+		hamtCID   = "bafybeidbclfqleg2uojchspzd4bob56dqetqjsj27gy2cq3klkkgxtpn4i"
+		hamtPath  = "/ipfs/" + rootCID + "/hamt/"
+		hamtRoots = rootCID + "," + hamtCID
```

Comment:
nit: this will not show issue with X-Roots-Path, we need `/hamt` that is more than one level dep, and load file that is in `/ipfs/bafybeidbcy4u6y55gsemlubd64zk53xoxs73ifd6rieejxcr7xy46mjvky/hamt/some-file` and then confirm the `X-Ipfs-Roots` includes only 3 CIDs for each path segment, and no internal ones from `/hamt/`.


There is some prior art fixture [here in gateway conformance](https://github.com/ipfs/gateway-conformance/blob/43c0842a1041a2748a74a196e370b84ef02b12a5/tests/trustless_gateway_car_test.go#L44) if that helps.

@hacdias  I don't want to block this PR on this, as it already brings a lot of other useful tests, so if you want fine to merge as-is and fill issue to fix this test in future PR.

Reply:
Ah, yes. I updated the test as I even used the same fixture. Will merge once CI is green.